src/Backups/RestoreCoordinationRemote.cpp

```diff
@@ -234,6 +235,46 @@ bool RestoreCoordinationRemote::acquireReplicatedSQLObjects(const String & loade
     return result;
 }
 
+bool RestoreCoordinationRemote::acquireInsertingDataForKeeperMap(const String & root_zk_path)
+{
+    bool result = false;
+    auto holder = with_retries.createRetriesControlHolder("acquireInsertingDataForKeeperMap");
+    holder.retries_ctl.retryLoop(
+        [&, &zk = holder.faulty_zookeeper]()
+        {
+            with_retries.renewZooKeeper(zk);
+
+            /// we need to remove leading '/' from root_zk_path
+            auto normalized_root_zk_path = std::string_view{root_zk_path}.substr(1);
+            std::string restore_lock_path = fs::path(zookeeper_path) / "keeper_map_tables" / normalized_root_zk_path / "restore_lock";
+            zk->createAncestors(restore_lock_path);
+            result = zk->tryCreate(restore_lock_path, "restorelock", zkutil::CreateMode::Persistent) == Coordination::Error::ZOK;
```

Comment:
That's not enough, because `tryCreate()` can possibly create a node and then fail and report failure. After that `holder.retries_ctl` will retry creating that node again and so it will decide that `InsertingDataForKeeperMap` is already acquired which is not true. Thus it's possible that `RESTORE` will succeed but leave a KeeperMap table empty.

To fix you can use as value `table_unique_id`:
```
result = zk->tryCreate(restore_lock_path, table_unique_id, zkutil::CreateMode::Persistent);
```
and if result is `ZNODEEXISTS` also check that its value is `table_unique_id`